clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerNetworkClient.java

```diff
@@ -191,26 +194,29 @@ public boolean poll(RequestFuture<?> future, long timeout) {
      * @throws WakeupException if {@link #wakeup()} is called from another thread
      */
     public void poll(long timeout) {
-        poll(timeout, time.milliseconds());
+        poll(timeout, time.milliseconds(), null);
     }
 
     /**
      * Poll for any network IO.
      * @param timeout timeout in milliseconds
      * @param now current time in milliseconds
      */
-    public void poll(long timeout, long now) {
+    public void poll(long timeout, long now, PollCondition pollCondition) {
         // there may be handlers which need to be invoked if we woke up the previous call to poll
         firePendingCompletedRequests();
 
         synchronized (this) {
             // send all the requests we can send now
             trySend(now);
 
-            // ensure we don't poll any longer than the deadline for
-            // the next scheduled task
-            client.poll(timeout, now);
-            now = time.milliseconds();
+            // check whether the poll is still needed by the caller. Note that if the expected completion
+            // condition becomes satisfied after the call to pollNeeded (because of a fired completion
+            // handler), the client will be woken up.
+            if (pollCondition == null || pollCondition.pollNeeded()) {
```

Comment:
We still do `checkDisconnects` and a few other operations even if we don't `poll`. Does that make sense? Thinking about this some more, if our concern is blocking on a `poll` because a background thread may have already completed the future, etc., would it make sense to do a `poll(0)` instead of no poll at all? I was thinking that it could be a bit more robust, but I may be missing something.